src/fs/cluster_allocator.cc

```diff
@@ -21,34 +21,98 @@
 
 #include "fs/cluster.hh"
 #include "fs/cluster_allocator.hh"
+#include "seastar/core/future.hh"
 
 #include <cassert>
 #include <optional>
 
-namespace seastar {
+namespace seastar::fs {
 
-namespace fs {
+namespace {
 
-cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters, std::deque<cluster_id_t> free_clusters)
-        : _allocated_clusters(std::move(allocated_clusters)), _free_clusters(std::move(free_clusters)) {}
-
-std::optional<cluster_id_t> cluster_allocator::alloc() {
-    if (_free_clusters.empty()) {
-        return std::nullopt;
+std::unordered_map<cluster_id_t, bool> create_allocated_clusters(const std::unordered_set<cluster_id_t>& allocated_clusters,
+        const circular_buffer<cluster_id_t>& free_clusters) {
+    std::unordered_map<cluster_id_t, bool> allocated_clusters_tmp;
+    for (auto& i : allocated_clusters) {
+        allocated_clusters_tmp.emplace(i, true);
+    }
+    for (auto& i : free_clusters) {
+        allocated_clusters_tmp.emplace(i, false);
     }
+    return allocated_clusters_tmp;
+}
+
+} // namespace
+
+cluster_allocator::cluster_allocator(std::unordered_set<cluster_id_t> allocated_clusters,
+        circular_buffer<cluster_id_t> free_clusters)
+        : _allocated_clusters(create_allocated_clusters(allocated_clusters, free_clusters))
+        , _free_clusters(std::move(free_clusters))
+        , _cluster_sem(_free_clusters.size()) {
+    _free_clusters.reserve(_free_clusters.size() + allocated_clusters.size());
+}
+
+void cluster_allocator::reset(std::unordered_set<cluster_id_t> allocated_clusters,
+        circular_buffer<cluster_id_t> free_clusters) {
+    assert((size_t)_cluster_sem.available_units() == _free_clusters.size());
+    assert(_cluster_sem.waiters() == 0);
+
+    free_clusters.reserve(free_clusters.size() + allocated_clusters.size());
+    _allocated_clusters = create_allocated_clusters(allocated_clusters, free_clusters);
+    _free_clusters = std::move(free_clusters);
+
+    _cluster_sem.consume(_cluster_sem.available_units());
+    _cluster_sem.signal(_free_clusters.size());
```

Comment:
maybe calling reset() from a constructor will be sufficient? I.e. constructor calls default constructor and then calls reset(), then you could inline create_allocated_clusters in the reset().